src/Chapter3.hs

```diff
@@ -480,6 +532,41 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Building = Church | Library
+
+{-
+params:	Int - population
+		Building - either church or library
+		String - name of castle
+		Bool - if wall is present
+-}
+data City
+	= SmallCity Int Building
+	| CityOfWizards Int Building String Bool
+
+buildCastle :: City -> String -> City
+buildCastle c name = case c of
+	SmallCity p b -> CityOfWizards p b name False
+	CityOfWizards p b _ w -> CityOfWizards p b name w
+
+buildWalls :: City -> City
+buildWalls c = case c of
+	SmallCity _ _ -> error "Build a castle before building walls"
```

Comment:
It's usually not a good idea to fail in runtime with `error`, because you can't easily catch such errors. The use case for `error` is rather something like `TODO` or to handle cases that are impossible to happen. But in this scenario, it's totally possible :heavy_check_mark: 

One of possible solutions to this problem is to return the original `city` without changes.

Alternatively, you can create a sum type to model all possible outcomes:

```haskell
data BuildWallsResult
    = NoCastle
    | NotEnoughPeople
    | NewCity City
```